_posts/2017-08-25-introducing-sandboxfs.md

```diff
@@ -3,40 +3,46 @@ layout: posts
+
+# Correct builds
+
 As our motto claims, *correctness* is an integral part of Bazel. To achieve
 correctness, builds must be *hermetic* and
 [*reproducible*](https://reproducible-builds.org/), which means that all
 actions invoked by Bazel should be run in a well-defined environment. In other
 words: we want actions to run within a sandbox.
 
-To better illustrate this point, let's consider an example: think of a C++
-`cc_library` target that specifies `foo.cc` as a source. This target contains a
+Let's consider an example. Think of a `cc_library` target that specifies
+`foo.cc` as a source. The `cc_library` rule contains a
 *compilation action* that turns the `foo.cc` source file into a `foo.o` object
-file. This action will run clang, will need to read `foo.cc`, will need to read
-the transitive closure of all headers required by `foo.cc`, will need to write
-`foo.o`, and may need to write some temporary files under `/tmp/`.
-
-Under these conditions, Bazel must ensure that the action only has read-only
-access to the input files we identified and write-only access to the output and
-temporary files we expect. The reason for this is because, unless these
-conditions are enforced, Bazel cannot know that the output file really came
-from those inputs: it may well be that the compiler went astray and read random
-files from the system, which means further builds would be inconsistent.
+file. This action will run clang, and it will need to:
+*  read `foo.cc`,
+*  read the transitive closure of all headers required by `foo.cc`
+*  write `foo.o`, and
+*  possibly write some temporary files under `/tmp/`.
+
+To be confident this build is correct, Bazel must ensure that the action has
+read-only access to the identified input files and write-only access
+to the output and temporary files we expect. Otherwise, Bazel cannot know if the
+compiler went astray and read random files from the system, making future builds
+inconsistent.
 
 # Current sandboxing techniques
 
 Today, Bazel uses different technologies to implement the sandboxing of
 actions. For example, on Linux, Bazel uses PID- and mount-namespaces, and on
-macOS uses symlinks and the `sandbox(7)` facility. (Note: sandboxing is
-disabled by default.)
+macOS Bazel uses symlinks and the `sandbox(7)` facility. Note that sandboxing
+is disabled by default.
 
-However, all these approaches have scalability and performance issues. The
-reason is that a typical build action requires hundreds, if not thousands, of
-files and directories to be mapped into the sandbox, and setting these up takes
-time. This is especially problematic on macOS: on this platform, Bazel must
+However, all these approaches have scalability and performance issues. A
+typical build action requires hundreds, if not thousands, of
+files and directories to be mapped into the sandbox. Setting these up takes
+time. On macOS this approach is especially problematic because Bazel must
```

Comment:
"On macOS this approach" -- is this correct or should this be "approaches"?

Reply:
I think singular is correct here. Approach refers to the creation of the many symlinks (or the mapping of the files/directories), but the creation is a singular action.

Reply:
cool, so there's just the one method / approach for macOS.